Zend/zend_compile.c

```diff
@@ -7189,12 +7189,71 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
+		if (ZEND_TYPE_FULL_MASK(ce->__unset->common.arg_info[-1].type) != MAY_BE_VOID) {
+			zend_error_noreturn(E_COMPILE_ERROR, "%s::__unset(): Return type must be void when declared", ZSTR_VAL(ce->name));
+		}
+	}
+
+	if (ce->__isset && ce->__isset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		 if (ZEND_TYPE_FULL_MASK(ce->__isset->common.arg_info[-1].type) != MAY_BE_BOOL) {
+			zend_error_noreturn(E_COMPILE_ERROR, "%s::__isset(): Return type must be bool when declared", ZSTR_VAL(ce->name));
+		 }
+	}
+
+    if (
+      ce->__debugInfo && ce->__debugInfo->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE
+      && !(ZEND_TYPE_FULL_MASK(ce->__debugInfo->common.arg_info[-1].type) & (MAY_BE_ARRAY | MAY_BE_NULL))
+    ) {
+      zend_error_noreturn(E_COMPILE_ERROR,
+        "Declaration of %s::__debugInfo() must be compatible with %s::__debugInfo(): ?array",
+        ZSTR_VAL(ce->name), ZSTR_VAL(ce->name));
+    }
+
+    if (zend_hash_str_exists(&ce->function_table, "__serialize", sizeof("__serialize") - 1)) {
+        zend_function *__serialize = zend_hash_str_find_ptr(&ce->function_table, "__serialize", sizeof("__serialize") - 1);
+
+        if (__serialize->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE && ZEND_TYPE_FULL_MASK(__serialize->common.arg_info[-1].type) != MAY_BE_ARRAY) {
+            zend_error_noreturn(E_COMPILE_ERROR, "%s::__serialize(): Return type must be array when declared", ZSTR_VAL(ce->name));
+        }
+    }
+
+    if (zend_hash_str_exists(&ce->function_table, "__unserialize", sizeof("__unserialize") - 1)) {
+        zend_function *__unserialize = zend_hash_str_find_ptr(&ce->function_table, "__unserialize", sizeof("__unserialize") - 1);
+
+        if (__unserialize->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE && ZEND_TYPE_FULL_MASK(__unserialize->common.arg_info[-1].type) != MAY_BE_VOID) {
+            zend_error_noreturn(E_COMPILE_ERROR, "%s::__unserialize(): Return type must be void when declared", ZSTR_VAL(ce->name));
+        }
+    }
+
+	if (zend_hash_str_exists(&ce->function_table, "__sleep", sizeof("__sleep") - 1)) {
+		zend_function *__sleep = zend_hash_str_find_ptr(&ce->function_table, "__sleep", sizeof("__sleep") - 1);
+
+		if (__sleep->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE && ZEND_TYPE_FULL_MASK(__sleep->common.arg_info[-1].type) != MAY_BE_ARRAY) {
+				zend_error_noreturn(E_COMPILE_ERROR, "%s::__sleep(): Return type must be array when declared", ZSTR_VAL(ce->name));
 		}
 	}
 
+    if (zend_hash_str_exists(&ce->function_table, "__wakeup", sizeof("__wakeup") - 1)) {
+        zend_function *__wakeup = zend_hash_str_find_ptr(&ce->function_table, "__wakeup", sizeof("__wakeup") - 1);
+
+        if (__wakeup->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE && ZEND_TYPE_FULL_MASK(__wakeup->common.arg_info[-1].type) != MAY_BE_VOID) {
+          zend_error_noreturn(E_COMPILE_ERROR, "%s::__wakeup(): Return type must be void when declared", ZSTR_VAL(ce->name));
+        }
+    }
```

Comment:
Why are argument types and return types checked in two different places?

Reply:
I've asked the same question here https://github.com/php/php-src/pull/4177#issuecomment-657248617.

Should we move all the checks to one place, or maybe even create a `zend_magic_methods` file?